File: common/src/main/java/io/github/techtastic/tisvs/util/RaycastHelper.java
package io.github.techtastic.tisvs.util;

import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.world.level.ClipContext;
import net.minecraft.world.phys.Vec3;
import org.valkyrienskies.mod.common.util.VectorConversionsMCKt;

public class RaycastHelper {
    public static ClipContext createContext(BlockPos pos, Direction direction, double distance) {
        Vec3 dir = VectorConversionsMCKt.toDoubles(direction.getNormal()).normalize();
        Vec3 origin = dir.add(VectorConversionsMCKt.toDoubles(pos));
        Vec3 end = origin.add(dir.x * distance, dir.y * distance, dir.z * distance);

        return new ClipContext(
                origin,
                end,
                ClipContext.Block.OUTLINE,
                ClipContext.Fluid.NONE,
                null
        );
    }
}

File: common/src/main/kotlin/io/github/techtastic/tisvs/util/RaycastHelper.kt
package io.github.techtastic.tisvs.util

import net.minecraft.core.BlockPos
import net.minecraft.core.Direction
import net.minecraft.world.level.ClipContext
import org.valkyrienskies.mod.common.util.toDoubles

object RaycastHelper {
    fun createContext(pos: BlockPos, direction: Direction, distance: Double): ClipContext {
        val dir = direction.normal.toDoubles().normalize()
        val origin = dir.add(pos.toDoubles())
        val end = origin.add(dir.x * distance, dir.y * distance, dir.z * distance)

        return ClipContext(
            origin,
            end,
            ClipContext.Block.OUTLINE,
            ClipContext.Fluid.NONE,
            null
        )
    }
}
